src/polyfills/custom-elements.js

```diff
@@ -0,0 +1,713 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @typedef {{
+ *   promise: !Promise<undefined>,
+ *   resolve: function(),
+ * }}
+ */
+let DeferredDef;
+
+/**
+ * @typedef {!Function}
+ */
+let CustomElementConstructorDef;
+
+/**
+ * @typedef {{
+ *  name: string,
+ *  ctor: !CustomElementConstructorDef,
+ * }}
+ */
+let CustomElementDef;
+
+/**
+ * Validates the custom element's name.
+ * This intentionally ignores "valid" higher Unicode Code Points.
+ * https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name
+ */
+const VALID_NAME = /^[a-z][a-z0-9._]*-[a-z0-9._-]*$/;
```

Comment:
I'm curious if this should be different for AMP's specific usecase. Could we only allow names matching  amp '-' [a-z] (PCENChar)*`? 

Anything else wouldn't be a valid AMP component.

Reply:
This would eliminate the need for an invalid names list of strings.

Reply:
I'd rather leave this generic since the plan is to extract this into a node module. We can add `amp-` prefix requirements at AMP's [`registerElement`](https://github.com/ampproject/amphtml/blob/53922a9c320b79ea049d3cd9efc07729c3c8f337/src/service/custom-element-registry.js#L136-L155) if we want.